File: app/src/main/java/com/mkulesh/onpc/iscp/messages/FirmwareUpdateMsg.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.jayway.jsonpath.JsonPath;
import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

/*
 * Firmware Update message
 */
public class FirmwareUpdateMsg extends ISCPMessage
{
    public final static String CODE = "UPD";

    public enum Status implements DcpStringParameterIf
    {
        NONE("N/A", "N/A", R.string.device_firmware_none),
        ACTUAL("FF", "update_none", R.string.device_firmware_actual),
        NEW_VERSION("00", "update_exist", R.string.device_firmware_new_version),
        NEW_VERSION_NORMAL("01", "N/A", R.string.device_firmware_new_version),
        NEW_VERSION_FORCE("02", "N/A", R.string.device_firmware_new_version),
        UPDATE_STARTED("Dxx-xx", "N/A", R.string.device_firmware_update_started),
        UPDATE_COMPLETE("CMP", "N/A", R.string.device_firmware_update_complete),
        NET("NET", "N/A", R.string.device_firmware_net);

        final String code, dcpCode;

        @StringRes
        final int descriptionId;

        Status(final String code, final String dcpCode, @StringRes final int descriptionId)
        {
            this.code = code;
            this.dcpCode = dcpCode;
            this.descriptionId = descriptionId;
        }

        public String getCode()
        {
            return code;
        }

        @NonNull
        public String getDcpCode()
        {
            return dcpCode;
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }
    }

    private final Status status;

    FirmwareUpdateMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        status = (Status) searchParameter(data, Status.values(), Status.NONE);
    }

    public FirmwareUpdateMsg(Status status)
    {
        super(0, null);
        this.status = status;
    }

    public Status getStatus()
    {
        return status;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + data + "; STATUS=" + status.getCode() + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return new EISCPMessage(CODE, status.getCode());
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }

    /*
     * Denon control protocol
     */
    private final static String HEOS_COMMAND = "player/check_update";

    @Nullable
    public static FirmwareUpdateMsg processHeosMessage(@NonNull final String command, @NonNull final String heosMsg)
    {
        if (HEOS_COMMAND.equals(command))
        {
            final Status s = (Status) searchDcpParameter(
                    JsonPath.read(heosMsg, "$.payload.update"), Status.values(), null);
            if (s != null)
            {
                return new FirmwareUpdateMsg(s);
            }
        }
        return null;
    }

    @Nullable
    @Override
    public String buildDcpMsg(boolean isQuery)
    {
        if (isQuery)
        {
            return "heos://" + HEOS_COMMAND + "?pid=" + DCP_HEOS_PID;
        }
        return null;
    }
}

File: lib/iscp/messages/FirmwareUpdateMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "../../constants/Strings.dart";
import "../DcpHeosMessage.dart";
import "../EISCPMessage.dart";
import "../ISCPMessage.dart";
import "EnumParameterMsg.dart";

enum FirmwareUpdate
{
    NONE,
    ACTUAL,
    NEW_VERSION,
    NEW_VERSION_NORMAL,
    NEW_VERSION_FORCE,
    UPDATE_STARTED,
    UPDATE_COMPLETE,
    NET
}

/*
 * Firmware Update message
 */
class FirmwareUpdateMsg extends EnumParameterMsg<FirmwareUpdate>
{
    static const String CODE = "UPD";

    static const ExtEnum<FirmwareUpdate> ValueEnum = ExtEnum<FirmwareUpdate>([
        EnumItem.code(FirmwareUpdate.NONE, "N/A", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_none, defValue: true),
        EnumItem.code(FirmwareUpdate.ACTUAL, "FF", dcpCode: "update_none",
            descrList: Strings.l_device_firmware_actual),
        EnumItem.code(FirmwareUpdate.NEW_VERSION, "00", dcpCode: "update_exist",
            descrList: Strings.l_device_firmware_new_version),
        EnumItem.code(FirmwareUpdate.NEW_VERSION_NORMAL, "01", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_new_version),
        EnumItem.code(FirmwareUpdate.NEW_VERSION_FORCE, "02", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_new_version),
        EnumItem.code(FirmwareUpdate.UPDATE_STARTED, "Dxx-xx", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_update_started),
        EnumItem.code(FirmwareUpdate.UPDATE_COMPLETE, "CMP", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_update_complete),
        EnumItem.code(FirmwareUpdate.NET, "NET", dcpCode: "N/A",
            descrList: Strings.l_device_firmware_net)
    ]);

    FirmwareUpdateMsg(EISCPMessage raw) : super(CODE, raw, ValueEnum);

    FirmwareUpdateMsg.output(FirmwareUpdate v) : super.output(CODE, v, ValueEnum);

    EnumItem<FirmwareUpdate> get getStatus
    => getValue;

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }

    /*
     * Denon control protocol
     */
    static const String _HEOS_COMMAND = "player/check_update";

    static FirmwareUpdateMsg? processHeosMessage(DcpHeosMessage jsonMsg)
    {
        if (_HEOS_COMMAND == jsonMsg.command)
        {
            final EnumItem<FirmwareUpdate>? s = ValueEnum.valueByDcpCode(jsonMsg.getString("payload.update"));
            return (s != null) ? FirmwareUpdateMsg.output(s.key) : null;
        }
        return null;
    }

    @override
    String? buildDcpMsg(bool isQuery)
    {
        if (isQuery)
        {
            return "heos://" + _HEOS_COMMAND + "?pid=" + ISCPMessage.DCP_HEOS_PID;
        }
        else if (getStatus.toString() == "NET")
        {
            return '<cmd id=\"3\"><name>SetUpdate</name><list><param name="\start\">1</param></list></cmd>';
        }
        return null;
    }
}
